plugins/inputs/azure_monitor/azure_monitor.go

```diff
@@ -0,0 +1,856 @@
+
+	bodyData, err := unmarshalJSON(body)
+
+	if err != nil {
+		return err
+	}
+
+	for _, value := range bodyData["value"].([]interface{}) {
+		resourceGroup := value.(map[string]interface{})["name"].(string)
+		resourceGroupTarget := NewResourceGroupTarget(resourceGroup, am.SubscriptionTargets)
+
+		am.ResourceGroupTargets = append(am.ResourceGroupTargets, resourceGroupTarget)
+	}
+
+	return nil
+}
+
+func (am *AzureMonitor) createResourceTargetsFromResourceGroupTargets() error {
+	var waitGroup sync.WaitGroup
+
+	errChan := make(chan error, 1)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	defer cancel()
+
+	for index, target := range am.ResourceGroupTargets {
+		am.Log.Info("Creating resource targets from resource group target #", index+1)
+		waitGroup.Add(1)
+
+		go func(target *ResourceGroupTarget) {
+			defer waitGroup.Done()
+
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			err := am.createResourceTargetFromResourceGroupTarget(target)
+
+			if err != nil {
+				select {
+				case errChan <- err:
+				default:
+				}
+
+				cancel()
+				return
+			}
+		}(target)
+	}
+
+	waitGroup.Wait()
+
+	select {
+	case err := <-errChan:
+		return err
+	default:
+	}
+
+	return nil
+}
+
+func (am *AzureMonitor) createResourceTargetFromResourceGroupTarget(target *ResourceGroupTarget) error {
+	apiURL := am.buildResourceGroupResourcesAPIURL(target)
+	body, err := am.getTargetResponseBody(apiURL)
+
+	if err != nil {
+		return err
+	}
+
+	bodyData, err := unmarshalJSON(body)
+
+	if err != nil {
+		return err
+	}
+
+	for _, value := range bodyData["value"].([]interface{}) {
+		fullResourceID := strings.Split(value.(map[string]interface{})["id"].(string), "/")
+		resourceID := strings.Join(fullResourceID[3:], "/")
+		resourceType := value.(map[string]interface{})["type"].(string)
+		resourceIndex := target.getResourceWithResourceTypeIndex(resourceType)
+
+		if resourceIndex == -1 {
+			continue
+		}
+
+		resourceTarget := NewResourceTarget(resourceID, target.Resources[resourceIndex].Metrics, target.Resources[resourceIndex].Aggregation)
+
+		am.ResourceTargets = append(am.ResourceTargets, resourceTarget)
+	}
+
+	return nil
+}
+
+func (am *AzureMonitor) getResourceTargetsMetrics() error {
+	var waitGroup sync.WaitGroup
+
+	errChan := make(chan error, 1)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	defer cancel()
+
+	for index, target := range am.ResourceTargets {
+		if len(target.Metrics) > 0 {
+			continue
+		}
+
+		am.Log.Info("Getting metrics for target #", index+1)
+		waitGroup.Add(1)
+
+		go func(target *ResourceTarget) {
+			defer waitGroup.Done()
+
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			err := am.getResourceTargetMetrics(target)
+
+			if err != nil {
+				select {
+				case errChan <- err:
+				default:
+				}
+
+				cancel()
+				return
+			}
+		}(target)
+	}
+
+	waitGroup.Wait()
+
+	select {
+	case err := <-errChan:
+		return err
+	default:
+	}
+
+	return nil
+}
```

Comment:
Same here for the parallel execution. If not absolutely necessary, we should do this sequentially to reduce complexity.